core/src/main/scala/org/apache/spark/eventhubs/PartitionsStatusTracker.scala

```diff
@@ -0,0 +1,379 @@
+
+
+      latestUpdatedBatch match {
+        case None => {
+          logDebug(s"No batch has ${PartitionsStatusTracker.enoughUpdatesCount} partitions with updates (enough updates), " +
+            s"so return None")
+          None
+        }
+        case Some(batch) => {
+          logDebug(s"Batch ${batch.batchId} is the latest batch with enough updates. Caculate and return its perforamnce.")
+          val performancePercentages = batch.getPerformancePercentages
+          PartitionsStatusTracker.throttlingStatusPlugin.foreach(
+            _.onPartitionsPerformanceStatusUpdate ( batch.batchId,
+                                                    batch.paritionsStatus.map(par => (par._1, par._2.batchSize))(breakOut),
+                                                    batch.paritionsStatus.map(par => (par._1, par._2.batchReceiveTimeInMillis))(breakOut),
+                                                    performancePercentages)
+          )
+          performancePercentages
+        }
+      }
+    }
+  }
+
+  /**
+   * Clean up the tracker. This will be called when the source has been stopped
+   */
+  def cleanUp() = {
+    batchesStatus.map(b => b._2.paritionsStatus.clear)
+    batchesStatus.clear
+    partitionSeqNoPairToBatchIdMap.clear
+  }
+
+  /**
+   * This methods i being used for testing
+   */
+  def batchIdsInTracker: scala.collection.Set[Long] = {
+    this.batchesStatus.keySet
+  }
+}
+
+object PartitionsStatusTracker {
+  private val _partitionsStatusTrackerInstance = new PartitionsStatusTracker
+  val BATCH_NOT_FOUND: Long = -1
+  val ACCEPTABLE_RECEIVE_TIME_PER_EVENT_MS: Double = 1.0
+  val ACCEPTABLE_DIFFERENCE_TIME_PER_EVENT_MS: Double = 1.0
+  var acceptableBatchReceiveTimeMS: Long = 500
+  var partitionsCount: Int = 1
+  var enoughUpdatesCount: Int = 1
+  var throttlingStatusPlugin: Option[ThrottlingStatusPlugin] = None
+  var defaultPartitionsPerformancePercentage: Option[Map[NameAndPartition, Double]] = None
+
+  def setDefaultValuesInTracker(numOfPartitions: Int, ehName: String, maxBatchReceiveTime: Long, throttlingSP: Option[ThrottlingStatusPlugin]) = {
+    partitionsCount = numOfPartitions
+    acceptableBatchReceiveTimeMS = maxBatchReceiveTime
+    enoughUpdatesCount = (partitionsCount/2) + 1
+    throttlingStatusPlugin = throttlingSP
+    defaultPartitionsPerformancePercentage =
+      Some((for (pid <- 0 until partitionsCount) yield (NameAndPartition(ehName, pid), 1.0))(breakOut))
+  }
+
+  private def partitionSeqNoKey(nAndP: NameAndPartition, seqNo: SequenceNumber): String =
+    s"(name=${nAndP.ehName},pid=${nAndP.partitionId},startSeqNo=$seqNo)".toLowerCase
+
+  def getPartitionStatusTracker: PartitionsStatusTracker = _partitionsStatusTrackerInstance
+}
+
+
+private[eventhubs] class BatchStatus(val batchId: Long,
+                                     val paritionsStatus: mutable.Map[NameAndPartition, PartitionStatus]) extends Logging{
```

Comment:
can you rename paritionsStatus -> paritionStatusList?

Reply:
Done